### ISSUE TITLE
status of snapshot volume object does not get updated

### ISSUE THREAD

**Issue by holgpar** [kubernetes-incubator/external-storage#764]
I am trying out the AWS example for the snapshot controller. The controller is able to talk to AWS and snapshots do get created.

However, the controller does not update the status part of the volumesnapshot resource. Also, I am seeing error message in the controller log:

```diff
-E0515 06:30:11.249368       1 snapshotter.go:303] No conditions for this snapshot yet.
I0515 06:30:11.255715       1 snapshotter.go:158] No VolumeSnapshotData objects found on the API server
I0515 06:30:11.255733       1 snapshotter.go:452] findSnapshot: snapshot snapshot-demo
-E0515 06:30:11.255746       1 snapshotter.go:459] Failed to get volume plugin. Snapshot snapshot-demo is malformed: invalid snapshot name
I0515 06:30:11.255782       1 snapshotter.go:443] syncSnapshot: Creating snapshot myns/snapshot-demo ...
I0515 06:30:11.255795       1 snapshotter.go:485] createSnapshot: Creating snapshot myns/snapshot-demo through the plugin ...
I0515 06:30:11.261896       1 snapshotter.go:491] createSnapshot: Creating metadata for snapshot myns/snapshot-demo.
I0515 06:30:11.261919       1 snapshotter.go:694] In updateVolumeSnapshotMetadata
I0515 06:30:11.263932       1 snapshotter.go:713] updateVolumeSnapshotMetadata: Metadata UID: cd72ffae-5783-11e8-bf9a-baa7acee7ea3 Metadata Name: snapshot-demo Metadata Namespace: myns Setting tags in Metadata Labels: map[string]string{"SnapshotMetadata-Timestamp":"1526365811263923331", "SnapshotMetadata-PVName":"pvc-1f741372-5780-11e8-bf9a-baa7acee7ea3"}.
```

**Comment by tsmetana** [comment]
Tested with the current master and everything seems to be working OK. Do you have the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY exported in the controller's environment? Can the controller access CA certificates? The most frequent problems are usually hinted by the controller logs on the start, make sure it doesn't complain about anything. If not, please post exact steps you used and yaml files so I can try to reproduce the problem.

**Comment by holgpar** [comment]
Hi tsmetana,

thanks for looking into this. The connection to aws seems to work fine, even the snapshots get created. To phrase the problem more precisely:  the VolumeSnapshotData object is not created.

Here is the complete log of the controller:

```
I0515 11:43:35.055494       1 aws.go:822] Building AWS cloudprovider
I0515 11:43:35.055596       1 aws.go:785] Zone not specified in configuration file; querying AWS metadata service
I0515 11:43:35.274723       1 tags.go:77] AWS cloud filtering on ClusterID: shoot-concise-eden
I0515 11:43:35.274817       1 snapshot-controller.go:122] Register cloudprovider aws
I0515 11:43:35.274834       1 snapshot-controller.go:94] starting snapshot controller
I0515 11:43:35.274895       1 snapshot-controller.go:167] Starting snapshot controller
I0515 11:43:35.274938       1 controller_utils.go:1019] Waiting for caches to sync for snapshot-controller controller
I0515 11:43:35.375137       1 controller_utils.go:1026] Caches are synced for snapshot-controller controller
I0515 11:48:08.853957       1 snapshot-controller.go:190] [CONTROLLER] OnAdd /apis/volumesnapshot.external-storage.k8s.io/v1/namespaces/myns/volumesnapshots/snapshot-demo, Snapshot &v1.VolumeSnapshot{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, Metadata:v1.ObjectMeta{Name:"snapshot-demo", GenerateName:"", Namespace:"myns", SelfLink:"/apis/volumesnapshot.external-storage.k8s.io/v1/namespaces/myns/volumesnapshots/snapshot-demo", UID:"d66844c3-5835-11e8-bf9a-baa7acee7ea3", ResourceVersion:"2424515", Generation:1, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63661981688, loc:(*time.Location)(0x280b920)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v1.VolumeSnapshotSpec{PersistentVolumeClaimName:"ebs-pvc", SnapshotDataName:""}, Status:v1.VolumeSnapshotStatus{CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Conditions:[]v1.VolumeSnapshotCondition(nil)}}
I0515 11:48:08.854064       1 desired_state_of_world.go:76] Adding new snapshot to desired state of world: myns/snapshot-demo
E0515 11:48:08.913969       1 snapshotter.go:303] No conditions for this snapshot yet.
I0515 11:48:08.920893       1 snapshotter.go:158] No VolumeSnapshotData objects found on the API server
I0515 11:48:08.920912       1 snapshotter.go:452] findSnapshot: snapshot snapshot-demo
E0515 11:48:08.920922       1 snapshotter.go:459] Failed to get volume plugin. Snapshot snapshot-demo is malformed: invalid snapshot name
I0515 11:48:08.920928       1 snapshotter.go:443] syncSnapshot: Creating snapshot myns/snapshot-demo ...
I0515 11:48:08.920933       1 snapshotter.go:485] createSnapshot: Creating snapshot myns/snapshot-demo through the plugin ...
I0515 11:48:08.936326       1 snapshotter.go:491] createSnapshot: Creating metadata for snapshot myns/snapshot-demo.
I0515 11:48:08.936348       1 snapshotter.go:694] In updateVolumeSnapshotMetadata
I0515 11:48:08.938494       1 snapshotter.go:713] updateVolumeSnapshotMetadata: Metadata UID: d66844c3-5835-11e8-bf9a-baa7acee7ea3 Metadata Name: snapshot-demo Metadata Namespace: myns Setting tags in Metadata Labels: map[string]string{"SnapshotMetadata-Timestamp":"1526384888938485220", "SnapshotMetadata-PVName":"pvc-1f741372-5780-11e8-bf9a-baa7acee7ea3"}.
I0515 11:48:08.941701       1 snapshot-controller.go:197] [CONTROLLER] OnUpdate oldObj: v1.VolumeSnapshotSpec{PersistentVolumeClaimName:"ebs-pvc", SnapshotDataName:""}
I0515 11:48:08.941782       1 snapshot-controller.go:198] [CONTROLLER] OnUpdate newObj: v1.VolumeSnapshotSpec{PersistentVolumeClaimName:"ebs-pvc", SnapshotDataName:""}
I0515 11:48:08.941931       1 snapshotter.go:734] updateVolumeSnapshotMetadata: returning cloudTags [map[string]string{"kubernetes.io/created-for/snapshot/namespace":"myns", "kubernetes.io/created-for/snapshot/name":"snapshot-demo", "kubernetes.io/created-for/snapshot/uid":"d66844c3-5835-11e8-bf9a-baa7acee7ea3", "kubernetes.io/created-for/snapshot/timestamp":"1526384888938485220"}]
```

I followed https://blog.jetstack.io/blog/volume-snapshotting/ to get it running. The yaml files I took over almost literally, only changing he image name. Is there maybe missning something on k8s side?

**Comment by tsmetana** [comment]
OK. I see the problem now. The first errors are actually harmless: due to the asynchronous nature of the controller it takes time for the actions to "settle" so the controller re-tries and spits error messages... I will take a look at improving the error handling.

As for the status updates: there really looks to be a bug. In my test the VolumeSnapshotData status conditions don't seem to be updated properly (stays "pending") even though the VolumeSnapshot itself is created and conditions were updated (I didn't notice at the first look). I'm not sure which commit were you using for your test but definitely the status handling is fishy and needs to be fixed.

**Comment by huang-jy** [comment]
I also encountered this same issue, but wasn't sure if it was something I was doing or not.

Point to note that the snapshot never even _started_ on the AWS console, even though the keys were being picked up by the controller and those keys have full/admin access.

**Comment by huang-jy** [comment]
Ah, that makes sense now, thanks @holgpar

**Comment by tsmetana** [comment]
Fixed by the PR #775.
